File: frontend/src/types/task.ts
export type TaskPriority = "low" | "medium" | "high";
export type TaskStatus = "pending" | "in_progress" | "done";

export const TASK_PRIORITIES: TaskPriority[] = ["low", "medium", "high"];
export const TASK_STATUSES: TaskStatus[] = ["pending", "in_progress", "done"];

export interface Task {
  id: string;
  title: string;
  description: string | null;
  priority: TaskPriority;
  status: TaskStatus;
  due_date: string | null;
  category: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskListResponse {
  items: Task[];
  total: number;
  page: number;
  page_size: number;
}

export interface TaskCreateInput {
  title: string;
  description?: string | null;
  priority?: TaskPriority;
  status?: TaskStatus;
  due_date?: string | null;
  category?: string | null;
}

export interface TaskUpdateInput {
  title?: string;
  description?: string | null;
  priority?: TaskPriority;
  status?: TaskStatus;
  due_date?: string | null;
  category?: string | null;
}

export type TaskSortField =
  | "created_at"
  | "updated_at"
  | "due_date"
  | "priority"
  | "title";
export type SortOrder = "asc" | "desc";

export interface TaskListParams {
  status?: TaskStatus;
  priority?: TaskPriority;
  category?: string;
  search?: string;
  due_before?: string;
  due_after?: string;
  page?: number;
  page_size?: number;
  sort_by?: TaskSortField;
  sort_order?: SortOrder;
}

File: backend/app/schemas/task.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.db.models.task import TaskPriority, TaskStatus


class TaskBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=10_000)
    priority: TaskPriority = TaskPriority.MEDIUM
    status: TaskStatus = TaskStatus.PENDING
    due_date: date | None = None
    category: str | None = Field(default=None, max_length=64)

    @field_validator("title")
    @classmethod
    def title_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Название не может быть пустым")
        return v.strip()


class TaskCreate(TaskBase):
    pass


class TaskUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=10_000)
    priority: TaskPriority | None = None
    status: TaskStatus | None = None
    due_date: date | None = None
    category: str | None = Field(default=None, max_length=64)

    @field_validator("title")
    @classmethod
    def title_not_blank(cls, v: str | None) -> str | None:
        if v is None:
            return v
        stripped = v.strip()
        if not stripped:
            raise ValueError("Название не может быть пустым")
        return stripped


class TaskRead(TaskBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class TaskListResponse(BaseModel):
    items: list[TaskRead]
    total: int
    page: int
    page_size: int
